### PR TITLE
first try of bdt using xgboost

### PR DESCRIPTION
**Pull request by gmilella12**
this is the first try of developing a bdt using xgboost ([documentation](https://xgboost.readthedocs.io/en/stable))
 
the core script is ```trainBDT.py```
which takes as inputs: 

- the input directory
- year of the samples
- the output directory (when not given, it creates the output in a subdirs in the input directory

it can be executed as follows: 
`python3 trainBDT.py --input_dir /nfs/dust/cms/user/gmilella/ttX_ntuplizer --year 2018 --output_dir ~/ttx3_analysis/CMSSW_11_1_7/src/PhysicsTools/NanoAODTools/scripts`

the script fetches the h5 file ```all_jets.h5``` that contains the dataframe of all jets already processed inclusively for signal and background, i.e. containing the inputs variables, the labels (0 -> bkg, 1-> sgn)**\***

this file is available here:
```/eos/user/g/gmilella/ttX_ntuplizer/total_jets_2018_h5 ```
```/nfs/dust/cms/user/gmilella/ttX_ntuplizer/total_jets_2018_h5```
if the file does not exist, the script will automatically process the individual files present in the sub-folders ```input_dir + sgn_2018_hotvr_h5 ``` and  ```input_dir + bkg_2018_hotvr_h5 ``` which are in my personal eos/nfs space...so we may need to change this last parsing options

* **N.B.: see next comment**

### CODE REVIEW

**Comment by gmilella12** [comment]
as a first simple try, I have separated signal and background such that:
- signal: 
all jets which contain **the full hadronic decay of a top from the resonance**. 
No distinction on the mass of the resonance is added to the dataframe

- background: 
all jets from pure **QCD processes** (not considering the ones which has a b quark inside for the moment). 
Therefore, only pure QCD samples have been studied for the time being

the results from xgboost is: 

```
### Train-test splitting
Split samples (65086, 5)
Train/test: (52068, 5)/(52068, 5)

XGBOOST PARAMETERS: {'n_estimators': 500, 'learning_rate': 0.1, 'max_depth': 4, 'min_child_weight': 0.01}

AUC: 0.9166627040515902 0.9166627040515902
Accuracy: 0.8321809940846585 0.8321809940846585

Bkg 30.0% -> Sigeff = 92.77091066344734, Thres = 0.3631054162979126
Bkg 20.0% -> Sigeff = 86.05715521280047, Thres = 0.509892463684082
Bkg 10.0% -> Sigeff = 73.10894086273379, Thres = 0.6811527013778687
Bkg 5.0% -> Sigeff = 60.36253558686799, Thres = 0.7865812182426453
Bkg 1.0% -> Sigeff = 37.367112328372194, Thres = 0.9114120006561279
```

**Comment by gmilella12** [comment]
- Commit "[0d379e4](https://github.com/ttXcubed/nanoAOD-tools/pull/20/commits/0d379e47a552f464caf5148e814bf49e16138eb5)" :
resampling procedure in bins of pT. The procedure is done in order to have similar pT distribution for signal and background jets. 

- Commit "[6ee7625](https://github.com/ttXcubed/nanoAOD-tools/pull/20/commits/6ee7625f80e1e5c6c1c6c04abeb0d30d49e06aab)":
template for condor submission for scanning different parameters of the BDT. 
The parameters are: 
```
n_trees_list = [500, 100]
learning_rate_list = [0.01, 0.03, 0.05, 0.08]
max_depth_list = [3, 5, 8]
min_child_weight_list = [.01, .05, .1, .5]
```
